editor/groups_editor.cpp

```diff
@@ -59,776 +61,809 @@ static bool can_edit(Node *p_node, String p_group) {
+	}
+};
 
-	if (!groups->is_anything_selected()) {
-		group_empty->hide();
+void GroupsEditor::_add_scene_group(const String &p_name) {
+	if (!node) {
 		return;
 	}
 
-	selected_group = groups->get_selected()->get_text(0);
-	_load_nodes(scene_tree->get_edited_scene_root());
-
-	group_empty->set_visible(!remove_node_root->get_first_child());
+	scene_groups[p_name] = true;
 }
 
-void GroupDialog::_load_nodes(Node *p_current) {
-	String item_name = p_current->get_name();
-	if (p_current != scene_tree->get_edited_scene_root()) {
-		item_name = String(p_current->get_parent()->get_name()) + "/" + item_name;
+void GroupsEditor::_remove_scene_group(const String &p_name) {
+	if (!node) {
+		return;
 	}
 
-	bool keep = true;
-	Node *root = scene_tree->get_edited_scene_root();
-	Node *owner = p_current->get_owner();
-	if (owner != root && p_current != root && !owner && !root->is_editable_instance(owner)) {
-		keep = false;
-	}
+	scene_groups.erase(p_name);
+	_remove_node_references(scene_root_node, p_name);
+}
 
-	TreeItem *node = nullptr;
-	NodePath path = scene_tree->get_edited_scene_root()->get_path_to(p_current);
-	if (keep && p_current->is_in_group(selected_group)) {
-		if (remove_filter->get_text().is_subsequence_ofn(String(p_current->get_name()))) {
-			node = nodes_to_remove->create_item(remove_node_root);
-			keep = true;
-		} else {
-			keep = false;
-		}
-	} else if (keep && add_filter->get_text().is_subsequence_ofn(String(p_current->get_name()))) {
-		node = nodes_to_add->create_item(add_node_root);
-		keep = true;
-	} else {
-		keep = false;
+void GroupsEditor::_rename_scene_group(const String &p_old_name, const String &p_new_name) {
+	if (!node) {
+		return;
 	}
 
-	if (keep) {
-		node->set_text(0, item_name);
-		node->set_metadata(0, path);
-		node->set_tooltip_text(0, path);
+	scene_groups[p_new_name] = scene_groups[p_old_name];
+	scene_groups.erase(p_old_name);
+	_rename_node_references(scene_root_node, p_old_name, p_new_name);
+}
 
-		Ref<Texture2D> icon = EditorNode::get_singleton()->get_object_icon(p_current, "Node");
-		node->set_icon(0, icon);
+void GroupsEditor::_remove_node_references(Node *p_node, const String &p_name) {
+	bool is_editable = can_edit(p_node, p_name);
 
-		if (!can_edit(p_current, selected_group)) {
-			node->set_selectable(0, false);
-			node->set_custom_color(0, groups->get_theme_color(SNAME("disabled_font_color"), SNAME("Editor")));
-		}
+	if (is_editable && p_node->is_in_group(p_name)) {
+		p_node->remove_from_group(p_name);
 	}
 
-	for (int i = 0; i < p_current->get_child_count(); i++) {
-		_load_nodes(p_current->get_child(i));
+	for (int i = 0; i < p_node->get_child_count(); i++) {
+		_remove_node_references(p_node->get_child(i), p_name);
 	}
 }
 
-void GroupDialog::_add_pressed() {
-	TreeItem *selected = nodes_to_add->get_next_selected(nullptr);
-
-	if (!selected) {
-		return;
+void GroupsEditor::_rename_node_references(Node *p_node, const String &p_old_name, const String &p_new_name) {
```

Comment:
I think you could make the GroupSettingsEditor method public and call it here instead of duplicating code.